contribute/design/1.x/csi/20190606-csi-volume-provisioning.md

```diff
@@ -0,0 +1,387 @@
+sustainable going forward
+- Building cstor volume controller deployment is currently done via go-templating
+which has become quite complex to maintain
+
+### Proposed Implementation
+Since OpenEBS itself runs on Kubernetes; limiting OpenEBS provisioning
+to be confined to CSI standards will prove difficult to develop and improve upon
+OpenEBS features.
+
+This proposal tries to adhere to CSI standards and make itself CSI compliant. 
+At the same time, this proposal lets OpenEBS embrace use of Kubernetes custom
+resources and custom controllers to provide storage for stateful applications.
+This also makes OpenEBS implementation idiomatic to Kubernetes practices which is
+implement features via Custom Resources.
+
+#### Volume Creation Workflow
+- CSI driver will handle CSI request for volume create
+- Kubernetes will have following as part of the infrastructure required to operate 
+OpenEBS CSI driver
+  - CStorVolumeConfigClass _(Kubernetes custom resource)_
+- CSI driver will read the request parameters and create following resources:
+  - CStorVolumeClaim _(Kubernetes custom resource)_
+- CStorVolumeClaim will be watched and reconciled by a dedicated controller
+
+#### Volume Deletion Workflow
+- CSI driver will handle CSI request for volume delete
+- CSI driver will read the request parameters and delete corresponding CStorVolume
+resource
+
+##### NOTES:
+- The resources owned by CStorVolume will be garbage collected
+
+### High Level Design
+This has the custom resources _(i.e. schemas)_ that ensure a smooth volume create
+& delete operations.
+
+#### CStorVolumeClaim -- new custom resource
+This resource is built & created by the CSI driver on a volume create request.
+This resource will be the trigger to get the desired i.e. requested cstor volume
+into the actual state in Kubernetes cluster. CStorVolumeClaim controller will 
+reconcile this claim into actual resources. CStorVolumeClaim controller will 
+operate from within the maya-apiserver pod.
+
+CStorVolumeClaim will be reconciled into following owned resources:
+- ConfigInjector _(Kubernetes custom resource)_
+- CStorVolume _(Kubernetes custom resource)_
+- CStor volume controller _(Kubernetes Deployment)_
+- CStor volume service _(Kubernetes Service)_
+- CStorVolumeReplica _(Kubernetes custom resource)_
+
+
+#### CStorVolumeConfigClass -- new custom resource
+This is a new Kubernetes custom resource that holds config information to be
+used by cstor volume claim controller while provisioning cstor volumes.
+
+NOTE: This resource kind does not have a controller. It holds the config 
+information to be utilized by cstor volume claim controller.
+
+Following is the proposed schema for `CStorVolumeConfigClass`:
+
+```yaml
+kind: CStorVolumeConfigClass
+metadata:
+  name: <some name -- can be same as storage class name>
+  namespace: <namespace of openebs i.e. openebs system namespace>
+spec:
+  controllerDeployment:
```

Comment:
`controller` is a generic word. Should we change this to `target`?

Reply:
Done